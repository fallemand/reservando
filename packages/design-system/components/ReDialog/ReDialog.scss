@import "../../styles/variables";

body.re-dialog-open {
  overflow: hidden;
}

.re-dialog {
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  position: fixed;
  z-index: $z-index-overlay;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  align-items: center;

  &--overlay {
    background: rgba(0, 0, 0, 0.5);
  }

  &__content {
    background: $white;
    position: relative;
    justify-self: flex-start;

    &--top {
      position: absolute;
      top: 0;
    }

    &--bottom {
      position: absolute;
      bottom: 0;
    }

    &--middle {
      margin: auto 0;
    }

    &--auto {
      width: auto;
    }

    &--fullwidth {
      width: 100%;
    }

    &--fullscreen {
      width: 100%;
      height: 100%;
    }
  }
}
