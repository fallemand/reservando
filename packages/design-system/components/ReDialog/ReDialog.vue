<template>
  <ReTransition v-show="show" name="fade">
    <div :class="['re-dialog', { 're-dialog--overlay': overlay }]" @click="handleOverlayClick">
      <ReTransition :name="transition">
        <div
          v-if="show"
          ref="content"
          :class="[
            're-dialog__content',
            `re-dialog__content--${size}`,
            `re-dialog__content--${verticalAlign}`,
            classDialog,
          ]"
        >
          <!-- @slot Default slot to pass any content -->
          <slot />
        </div>
      </ReTransition>
    </div>
  </ReTransition>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";
import ReTransition from "../ReTransition/ReTransition.vue";

export const MODAL_OPEN = "re-dialog-open";

/**
 * Generic pop up dialog to generate components like `ReModal`, `ReBottomSheet`, etc <br />
 * It handles the overlay, transitions, aligns, close with ESC, etc
 */
const ReDialog = defineComponent({
  name: "ReDialog",
  components: { ReTransition },
  props: {
    /** Size for the modal */
    size: {
      default: "auto",
      type: String as PropType<"auto" | "fullwidth" | "fullscreen">,
      validator: (value: string): boolean => ["auto", "fullwidth", "fullscreen"].includes(value),
    },
    /** Show overlay in the background */
    overlay: {
      type: Boolean,
      default: true,
    },
    /** Close dialog when clicking the background overlay */
    overlayClose: {
      type: Boolean,
      default: true,
    },
    /** Type of transition */
    transition: {
      type: String as PropType<"fade" | "expand" | "slide-bottom" | "slide-right">,
      default: "fade",
      validator: (value: string): boolean =>
        ["fade", "expand", "slide-bottom", "slide-right"].includes(value),
    },
    /** Vertical align of the dialog */
    verticalAlign: {
      type: String as PropType<"middle" | "bottom" | "top">,
      default: "middle",
      validator: (value: string): boolean => ["middle", "bottom", "top"].includes(value),
    },
    /** CSS class to add to the content div */
    classDialog: {
      type: [String, Array, Object],
      default: null,
    },
  },
  data() {
    return {
      show: false,
    };
  },
  mounted() {
    this.show = true;
    document.body.classList.add(MODAL_OPEN);
    document.body.addEventListener("keydown", this.onEsc);
  },
  unmounted() {
    document.body.classList.remove(MODAL_OPEN);
    document.body.removeEventListener("keydown", this.onEsc);
  },
  methods: {
    handleOverlayClick($evt: Event): void {
      const content = this.$refs.content as HTMLInputElement;
      if (this.overlayClose && content && !content.contains($evt.target as Node)) {
        $evt.stopPropagation();
        this.close($evt);
      }
    },
    onEsc($evt: KeyboardEvent): void {
      const escKeyCode = 27;
      if ($evt.keyCode === escKeyCode) {
        this.close($evt);
      }
    },
    async close($evt?: Event): Promise<void> {
      this.show = false;
      await this.$nextTick();
      /** `($evt: Event)` */
      setTimeout(() => {
        this.$emit("close", $evt);
      }, 300);
    },
  },
});

export default ReDialog;
</script>

<style lang="scss" src="./ReDialog.scss" />
