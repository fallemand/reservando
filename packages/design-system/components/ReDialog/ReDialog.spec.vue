<script lang="ts">
import { shallowMount, mount } from "@vue/test-utils";
import ReDialog from "./ReDialog.vue";

describe("ReDialog component", () => {
  it("renders component", async () => {
    const wrapper = shallowMount(ReDialog);
    // We need to wait for mounted() to execute, so show === true
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".re-dialog").exists()).toBeTruthy();
  });

  it("render with default props", async () => {
    const wrapper = shallowMount(ReDialog, {
      slots: {
        default: "__CONTENT__",
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("render with custom props", async () => {
    const wrapper = shallowMount(ReDialog, {
      propsData: {
        verticalAlign: "top",
        size: "fullscreen",
        overlay: false,
        transition: "slide-right",
        classDialog: "__CONTENT-CLASS__",
      },
      slots: {
        default: "__CONTENT__",
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should emit close", async () => {
    const wrapper = shallowMount(ReDialog);
    await wrapper.vm.$nextTick();
    const element = wrapper.find(".re-dialog");

    element.trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().close).toBeTruthy();
  });

  it("should have modal slot content", async () => {
    const wrapper = shallowMount(ReDialog, {
      slots: {
        default: "<p>test slot</p>",
      },
    });
    await wrapper.vm.$nextTick();
    const content = wrapper.find(".re-dialog__content");
    expect(content.text()).toEqual("test slot");
  });

  it("should close modal when esc pressed", async () => {
    const div = document.createElement("div");
    div.id = "root";
    document.body.appendChild(div);

    const wrapper = mount(ReDialog, {
      attachTo: div,
    });
    await wrapper.vm.$nextTick();

    wrapper.trigger("keydown", { keyCode: 27 });
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().close).toBeTruthy();
  });
});
</script>
