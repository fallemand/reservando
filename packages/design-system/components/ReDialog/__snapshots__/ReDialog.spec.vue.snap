// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReDialog component render with custom props 1`] = `
<re-transition-stub
  style=""
  name="fade"
/>
`;

exports[`ReDialog component render with default props 1`] = `
<re-transition-stub
  style=""
  name="fade"
/>
`;
